### PR TITLE
ENG-534: client usage updater

### PR DESCRIPTION
**Pull request by efiShtain**
Added an optional (configurable) updater that calls a predefined url (currently localhost) on predefined interval (currently once every 24 hours) and pass some usage data, on failure it doesn't do anything since this is just a nice to have feature

```go
type updateRequest struct {
	InstallationID string `json:"installationID"`
	Version        string `json:"version"`
	Commit         string `json:"commit"`
	OS             string `json:"os"`
	Arch           string `json:"arch"`
}
```

Matching PR to collect the data and change the endpoint will be added later

### CODE REVIEW

File: internal/backend/usageupdater/dataposter.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+func (p poster) post(data []byte) {
+	req, err := http.NewRequest("POST", p.endpoint, bytes.NewReader(data))
+	if err != nil {
+		return
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := p.client.Do(req)
+	if err != nil {
+		return
```
Comment (ig-ra) [inline internal/backend/usageupdater/dataposter.go]: log? here and below? at least in debug?

File: internal/backend/usageupdater/dataposter.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	endpoint string
+	client   http.Client
+}
+
+func newPoster(endpoint string) poster {
+	tr := &http.Transport{
+		DisableKeepAlives:   true,
+		MaxIdleConnsPerHost: -1,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
```
Comment (ig-ra) [inline internal/backend/usageupdater/dataposter.go]: dummy implementation for now?

File: internal/backend/usageupdater/usagedata.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	installationIDFile := filepath.Join(xdg.ConfigHomeDir(), "installID")
+
+	if _, err := os.Stat(installationIDFile); os.IsNotExist(err) {
+		id := uuid.New()
+		if err := os.WriteFile(installationIDFile, []byte(id.String()), 0600); err != nil {
+			return nil, err
+		}
+	}
+
+	data, err := os.ReadFile(installationIDFile)
```
Comment (ig-ra) [inline internal/backend/usageupdater/usagedata.go]: maybe if we fail to read and parse we need to recreate file?

**Review by ig-ra — CHANGES_REQUESTED** [review]

File: internal/backend/usageupdater/usagedata.go (hunk 3 above)
Comment (efiShtain) [reply internal/backend/usageupdater/usagedata.go]: IDK, 
sounds like a valid idea
I'll add it

File: internal/backend/usageupdater/dataposter.go (hunk 2 above)
Comment (efiShtain) [reply internal/backend/usageupdater/dataposter.go]: nope, we just don't worry about it

File: internal/backend/usageupdater/dataposter.go (hunk 1 above)
Comment (efiShtain) [reply internal/backend/usageupdater/dataposter.go]: we don't really have anything to do with the log, nor do we care if it fails
but I'll some logs in debug

File: internal/backend/basesvc/svc.go [hunk 4]
```diff
@@ -124,6 +125,11 @@ func makeFxOpts(cfg *Config, opts RunOptions) []fx.Option {
 		Component("triggers", configset.Empty, fx.Provide(triggers.New)),
 		Component("oauth", configset.Empty, fx.Provide(oauth.New)),
 		Component("runtimes", configset.Empty, fx.Provide(runtimes.New)),
+		Component("usageupdater", usageupdater.Configs, fx.Provide(usageupdater.New),
```
Comment (itayd) [inline internal/backend/basesvc/svc.go]: usagereporter?

File: internal/backend/usageupdater/config.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	Enabled  bool          `koanf:"enabled"`
+	Endpoint string        `koanf:"endpoint"`
+	Interval time.Duration `koadnf:"interval_seconds"`
+}
+
+var (
+	Configs = configset.Set[Config]{
+		Default: &Config{
+			Enabled:  true,
+			Endpoint: "http://localhost:9980",
```
Comment (itayd) [inline internal/backend/usageupdater/config.go]: i would put in dev for now, and disable default until we have the real thing

File: internal/backend/usageupdater/dataposter.go (hunk 2 above)
Comment (itayd) [reply internal/backend/usageupdater/dataposter.go]: should we?

File: internal/backend/usageupdater/dataposter.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := p.client.Do(req)
+	if err != nil {
+		return
+	}
+	if resp.StatusCode != 200 {
+		return
+	}
+	defer resp.Body.Close()
```
Comment (itayd) [inline internal/backend/usageupdater/dataposter.go]: defer on the last line is useless, just close it. i think you need to close in any status code?

File: internal/backend/usageupdater/usagedata.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	"time"
+
+	"github.com/google/uuid"
+	"go.autokitteh.dev/autokitteh/internal/version"
+	"go.autokitteh.dev/autokitteh/internal/xdg"
+	"go.uber.org/zap"
+)
+
+type updateRequest struct {
+	InstallationID string `json:"installationID"`
```
Comment (itayd) [inline internal/backend/usageupdater/usagedata.go]: installation_id

File: internal/backend/usageupdater/usagedata.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	shutdownChan   chan struct{}
+	poster         poster
+}
+
+func New(z *zap.Logger, cfg *Config) (UsageUpdater, error) {
+	if !cfg.Enabled {
+		return &nopUpdater{}, nil
+	}
+
+	installationIDFile := filepath.Join(xdg.ConfigHomeDir(), "installID")
```
Comment (itayd) [inline internal/backend/usageupdater/usagedata.go:41]: where is this file coming from? how do we generate it?

File: internal/backend/usageupdater/usagedata.go (hunk 3 above)
Comment (itayd) [reply internal/backend/usageupdater/usagedata.go]: so this will replace lines 43-48?

File: internal/backend/usageupdater/usagedata.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+			return nil, err
+		}
+	}
+
+	data, err := os.ReadFile(installationIDFile)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := uuid.Parse(string(data))
```
Comment (itayd) [inline internal/backend/usageupdater/usagedata.go]: if this fails we'll need to recreate it as well?

File: internal/backend/usageupdater/usagedata.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		Arch:           runtime.GOARCH,
+	}
+
+	data, err := json.Marshal(r)
+	if err != nil {
+		return
+	}
+
+	d.poster.post(data)
+
```
Comment (itayd) [inline internal/backend/usageupdater/usagedata.go]: kill

File: internal/backend/usageupdater/usagedata.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+func (d *usageUpdater) Start() {
+	go func() {
+		timer := time.NewTicker(d.updateInterval)
```
Comment (itayd) [inline internal/backend/usageupdater/usagedata.go]: you don't need this, just replace timer.C with time.After

File: internal/backend/usageupdater/usagedata.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+		defer timer.Stop()
+
+		d.update()
+		for {
+			select {
+			case <-d.shutdownChan:
+				return
+			case <-timer.C:
+				d.update()
```
Comment (itayd) [inline internal/backend/usageupdater/usagedata.go]: can you add a debug log when we update?

File: internal/backend/usageupdater/dataposter.go (hunk 2 above)
Comment (efiShtain) [reply internal/backend/usageupdater/dataposter.go]: not sure we care about the certificate here
I can remove the tis skip verify part
took reference from temporal code, they do the same

File: internal/backend/usageupdater/usagedata.go (hunk 3 above)
Comment (efiShtain) [reply internal/backend/usageupdater/usagedata.go]: Changed the logic a bit 
please take a look again

File: internal/backend/basesvc/svc.go (hunk 4 above)
Comment (efiShtain) [reply internal/backend/basesvc/svc.go]: agree, better name

File: internal/backend/usageupdater/config.go (hunk 5 above)
Comment (efiShtain) [reply internal/backend/usageupdater/config.go]: done

File: internal/backend/usageupdater/dataposter.go (hunk 6 above)
Comment (efiShtain) [reply internal/backend/usageupdater/dataposter.go]: correct
moved it up two lines

File: internal/backend/usageupdater/usagedata.go (hunk 7 above)
Comment (efiShtain) [reply internal/backend/usageupdater/usagedata.go]: done

File: internal/backend/usageupdater/usagedata.go (hunk 10 above)
Comment (efiShtain) [reply internal/backend/usageupdater/usagedata.go]: ?

File: internal/backend/usageupdater/usagedata.go (hunk 12 above)
Comment (efiShtain) [reply internal/backend/usageupdater/usagedata.go]: sure

File: internal/backend/usageupdater/usagedata.go (hunk 9 above)
Comment (efiShtain) [reply internal/backend/usageupdater/usagedata.go]: good point, 
tried to resolve it
make the code a bit more "error handli"
let me know what you think

File: internal/backend/usageupdater/usagedata.go (hunk 11 above)
Comment (efiShtain) [reply internal/backend/usageupdater/usagedata.go]: nice, updated

File: internal/backend/usagereporter/usagedata.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/google/uuid"
+	"go.autokitteh.dev/autokitteh/internal/version"
+	"go.autokitteh.dev/autokitteh/internal/xdg"
+	"go.uber.org/zap"
```
Comment (daabr) [inline internal/backend/usagereporter/usagedata.go]:
```suggestion
	"github.com/google/uuid"
	"go.uber.org/zap"

	"go.autokitteh.dev/autokitteh/internal/version"
	"go.autokitteh.dev/autokitteh/internal/xdg"
```

File: internal/backend/usagereporter/usagedata.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+	return string(data), nil
+}
+
+func New(z *zap.Logger, cfg *Config) (UsageReporter, error) {
+	if !cfg.Enabled {
+		return &nopUpdater{}, nil
+	}
+
+	installationIDFile := filepath.Join(xdg.ConfigHomeDir(), "installID")
```
Comment (daabr) [inline internal/backend/usagereporter/usagedata.go]: I'm not sure ConfigHomeDir is the right choice - maybe DataHomeDir because it's not exactly "configurable"?

But it's up to you.

File: internal/backend/usagereporter/usagedata.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+		InstallationID: d.installationID.String(),
+		Version:        version.Version,
+		Commit:         version.Commit,
+		OS:             runtime.GOOS,
+		Arch:           runtime.GOARCH,
+	}
+
+	data, err := json.Marshal(r)
+	if err != nil {
+		d.logger.Debug("faild report usage data", zap.Error(err))
```
Comment (daabr) [inline internal/backend/usagereporter/usagedata.go]: Nit: Should be "failed to report..."

Slightly better text, if you want: "usage data reporting error/failed" (and in line 121 you can use "usage data reporting success")

File: internal/backend/usagereporter/usagedata.go [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+		installationID: id,
+		updateInterval: cfg.Interval,
+		shutdownChan:   make(chan struct{}),
+		endpoint:       cfg.Endpoint,
+		logger:         z,
+	}, nil
+}
+
+func (d *usageReporter) report() {
+	r := reportRequest{
```
Comment (daabr) [inline internal/backend/usagereporter/usagedata.go:106]:
```suggestion
func (d *usageReporter) report() {
	// https://github.com/search?q=repo%3Aautokitteh%2Fautokitteh%20ldflags&type=code
	// https://github.com/autokitteh/homebrew-tap/blob/main/Formula/autokitteh.rb#L14
	r := reportRequest{
```

File: internal/backend/usagereporter/usagedata.go [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+type usageReporter struct {
+	installationID uuid.UUID
+	updateInterval time.Duration
+	shutdownChan   chan struct{}
+	endpoint       string
+	logger         *zap.Logger
+}
+
+func generateInstallIDFile(path string) error {
+	id := uuid.New()
```
Comment (daabr) [inline internal/backend/usagereporter/usagedata.go:39]: Question: did you intentionally choose UUIDv4? Any opinion about which UUID is better for this purpose - v4 (random) or v7 (timestamp + random)?

File: internal/backend/usagereporter/usagedata.go [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+	if err := os.WriteFile(path, []byte(id.String()), 0600); err != nil {
```
Comment (daabr) [inline internal/backend/usagereporter/usagedata.go]:
```suggestion
	if err := os.WriteFile(path, []byte(id.String()), 0o600); err != nil {
```

File: internal/backend/usagereporter/usagedata.go [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+	// if there is an error and the file exists, try to remove it and fail if cant
+	// generate a new file
+	// try read the data again
+	data, err := os.ReadFile(installationIDFile)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			if err := os.Remove(installationIDFile); err != nil {
+				return "", err
+			}
+		}
```
Comment (daabr) [inline internal/backend/usagereporter/usagedata.go:57]: Why is this needed? According to https://pkg.go.dev/os#WriteFile WriteFile() rewrites the file if it exists, no?

**Review by daabr — APPROVED** [review]

File: internal/backend/usagereporter/usagedata.go (hunk 19 above)
Comment (efiShtain) [reply internal/backend/usagereporter/usagedata.go:57]: good point, removed

File: internal/backend/usagereporter/usagedata.go (hunk 18 above)
Comment (efiShtain) [reply internal/backend/usagereporter/usagedata.go]: done

File: internal/backend/usagereporter/usagedata.go (hunk 17 above)
Comment (efiShtain) [reply internal/backend/usagereporter/usagedata.go:39]: doesn't really matter, can also be shortid or something, since each machine can have different clock, we won't really get anything from the timestamp
it just needs to be random

File: internal/backend/usagereporter/usagedata.go (hunk 16 above)
Comment (efiShtain) [reply internal/backend/usagereporter/usagedata.go:106]: not sure I understand

File: internal/backend/usagereporter/usagedata.go (hunk 15 above)
Comment (efiShtain) [reply internal/backend/usagereporter/usagedata.go]: updated some

File: internal/backend/usagereporter/usagedata.go (hunk 14 above)
Comment (efiShtain) [reply internal/backend/usagereporter/usagedata.go]: changed

File: internal/backend/usagereporter/usagedata.go (hunk 16 above)
Comment (daabr) [reply internal/backend/usagereporter/usagedata.go:106]: I added comments that are references to the places in our codebase where we set version values. This may be relevant for future analysis of the data.

If it was all inside the autokitteh repo I wouldn't bother, but it's useful to remember that if AK is installed by brew its version looks slightly different.

File: internal/backend/usagereporter/usagedata.go (hunk 17 above)
Comment (daabr) [reply internal/backend/usagereporter/usagedata.go:39]: sgtm

File: internal/backend/usageupdater/dataposter.go (hunk 2 above)
Comment (itayd) [reply internal/backend/usageupdater/dataposter.go]: ok

File: internal/backend/usageupdater/usagedata.go (hunk 10 above)
Comment (itayd) [reply internal/backend/usageupdater/usagedata.go]: empty line

File: internal/backend/usageupdater/usagedata.go (hunk 3 above)
Comment (daabr) [reply internal/backend/usageupdater/usagedata.go]: ack

**Review by daabr — APPROVED** [review]

**Review by ig-ra — APPROVED** [review]
